### PR TITLE
PUC update to include Native Phase 2 spec:

### PR DESCRIPTION
**Pull request by mmoschovas**
1) Added renderNativeAd function to nativeTrackerManager.js
2) renderNativeAd checks for existence of adId and performs the following:
   - Checks for rendererUrl and preloads if exists
   - Uses prexisting loadAssets function with new callback
   - Fires Impression tracker and loads click tracker
3) LoadAssets update:
   - Checks for existence of assetsToReplace and adds each asset to placeholder
   - If no placeholders exist, checks for requestAllAssets flag to fire new requestAllAssets function
4) requestAllAssets will postmessage Prebid using the action allAssetRequest instead of assetRequest
5) Replace assets will now use added logic:
   - If rendererUrl existence. If exists in adServer (preloaded) then check for renderAd availability on the window. If none exists, set load event on the script tag using the id 'pb-native-renderer'. If url sent as asset from Prebid, append script to window and attach load event. When script loaded and renderAd available, fire function to retrieve HTML. This inserted into innerHtml
   - Else if adTemplate exists, replace assets and insert into innerHtml
   - Else run existing replace method
6) replace function updated to include check for new native to replace keys which are in the format of ##KEY##

nativeAssetManager_spec and nativeTrackerManager_spec both updated to include additional tests

nativeTracker.js updated to include renderNativeAd function definition on window object

### CODE REVIEW

File: src/nativeTrackerManager.js
Comment (wqi1972) [inline src/nativeTrackerManager.js:149]: Suggest move `renderNativeAd` to a new class instead of add into `newNativeTrackerManager`. In long run, `newNativeTrackerManager.startTrackers` will get deprecated. If we separate code `renderNativeAd` from `newNativeTrackerManager` now, it will be much easier when that happens.

File: src/nativeTrackerManager.js
Comment (wqi1972) [inline src/nativeTrackerManager.js:79]: We suppose set `publisherDomain` value inside `renderNativeAd` once, instead of set it before each tracker call.

**Comment by wqi1972** [comment]
I don't see code inside `renderNativeAd` take care of request for Mobile App as `startTrackers` does. I don't have experience on native ad for Mobile App, need someone has that experience double check it.

**Comment by wqi1972** [comment]
We need logic to take care `````` and `````` as https://github.com/prebid/prebid-universal-creative/pull/60/files does.
This way will avoid 400 error message for icon and hero image.

File: src/nativeTrackerManager.js
Comment (mmoschovas) [reply src/nativeTrackerManager.js]: renderAd code separated into new nativeRenderManager

File: src/nativeTrackerManager.js
Comment (mmoschovas) [reply src/nativeTrackerManager.js]: Corrected the above to one definition in renderNativeAd as opposed to the tracker definitions

File: src/nativeAssetManager.js
Comment (wqi1972) [inline src/nativeAssetManager.js:183]: I think we suppose check `requestAllAssets` first. If it is true, then `requestAllAssets`. Only when `requestAllAssets` is not true, then `requestAssets` for  `placeholders`.

File: src/nativeTrackerManager.js
Comment (wqi1972) [reply src/nativeTrackerManager.js]: Good.

File: src/nativeAssetManager.js
Comment (mmoschovas) [reply src/nativeAssetManager.js:183]: Can you describe the use case for this? I would think that if placeholders are detected/defined, then requesting all assets should not be a requirement as the needed assets would be known.

Comment (wqi1972) [reply src/nativeAssetManager.js:183]: 2 reasons:
1) for Use Case 1 - “Render Replacements”, and Use Case 2 - “Ad Template as String”, these 2 use cases both have placementholder token inside html. pbNativeTagData.requestAllAssets = true; will be useless if the code check placementholders exist first;
2) for Use Case 1 - “Render Replacements”, for html `icon`, there suppose have some code which get asset of `##hb_native_icon##`, and set src of the image be the value of `##hb_native_icon##`. But inside html, there is no placeholder for `##hb_native_icon##`, so the asset won't get pulled in existing logic.

Comment (wqi1972) [reply src/nativeAssetManager.js:183]: To make it easier, I will create a PR against your branch, so you will understand what I mean.

Comment (wqi1972) [reply src/nativeAssetManager.js:183]: unfortunately my account can't push commit into git@github.com:mmoschovas/prebid-universal-creative.git repo, even inside the my own branch.
Here is the code change I made, I change function fireNativeCallback() to:
```
  function fireNativeCallback(){
    const adElements = findAdElements(AD_ANCHOR_CLASS_NAME);
    for (let i = 0; i < adElements.length; i++) {
      adElements[i].addEventListener('click', function(event) {
        loadClickTrackers(event, window.pbNativeData.adId);
      }, true);
    }
    const imgElements = win.document.getElementsByTagName("img");
    for (let i = 0; i < imgElements.length; i++) {
      if (imgElements[i].dataset && imgElements[i].dataset.src) {
        imgElements[i].src = imgElements[i].dataset.src;
      }
    }
  }
```
with this change, then it will support  tag with following format
```
<img data-src="##hb_native_image##" alt="" />
<img data-src="##hb_native_icon##" alt="logo" />
```
This will avoid 400 error if we directly set img.src inside html template.

Comment (wqi1972) [reply src/nativeAssetManager.js:183]: Notice, this is a different solution than using ``, I think it is better and easier.
And with this solution, it doesn't need `pbNativeTagData.requestAllAssets = true;` for Use Case 1 - “Render Replacements” anymore, since all asset placementholder exists in GAM template html.

Comment (wqi1972) [reply src/nativeAssetManager.js:183]: One more notice, you need to adjust `renderingMocks` inside nativeRenderManager_spec.js to 
```
const renderingMocks = {
  getWindowObject: function() {
    return {
      document: {
        getElementsByTagName: function() {
          return [];
        }
      },
      parent: {
        postMessage: sinon.spy()
      }
    }
  }
};
```

**Comment by robertrmartinez** [comment]
Hey all, I am setting up some GAM line items today to help make testing local PUC changes easier.

I'll share example test pages once I am done today / tomorrow.

I would like to finalize the native stuff and merge both these for next weeks release!

Thanks again.

**Comment by wqi1972** [comment]
I've tested on my local with following test cases:
1) New PBJS + PUC code with existing configure and templates, (using native key/value). Confirmed it is backward compatible; 
2) New PBJS + PUC code + sendTargetingKeys: false + template with macro ##native_<asset name>##. Confirmed  Use Case 1 - “Render Replacements” is working. The only issue the  tag temporarily raises 400 error. And with my suggestion provided above, that 400 error will get solved.

File: dist/load-cookie.html
Comment (robertrmartinez) [inline dist/load-cookie.html:1]: I think this should not be checked in correct?

Comment (robertrmartinez) [reply dist/load-cookie.html:1]: Same with the .DS_Store binary.

File: src/nativeAssetManager.js
Comment (robertrmartinez) [inline src/nativeAssetManager.js:277]: some lint issues (we do not have automatic linter on puc code yet)

Maybe run the entire code on your local with your eslint plugin and do the fixes?

Comment (robertrmartinez) [inline src/nativeAssetManager.js:277]: @jaiminpanchal27 When the rendererUrl is loaded, is the expected output that a renderAd function is placed on the win. 

If you could review this use case a little more in depth I would appreciate it since it was spec'd by you.

File: src/nativeRenderManager.js
Comment (robertrmartinez) [inline src/nativeRenderManager.js:70]: some more lint issues

File: src/nativeTrackerManager.js
Comment (robertrmartinez) [inline src/nativeTrackerManager.js:13]: extra new line?

File: test/spec/nativeTrackerManager_spec.js
Comment (robertrmartinez) [inline test/spec/nativeTrackerManager_spec.js:98]: what is the reasoning behind this being commented out?

File: src/nativeTrackerManager.js
Comment (mmoschovas) [reply src/nativeTrackerManager.js:13]: Removing this. I believe I originally added logic that I later removed.

File: test/spec/nativeTrackerManager_spec.js
Comment (mmoschovas) [reply test/spec/nativeTrackerManager_spec.js:98]: I forgot to remove this test. Originally the new rendering code lived within trackerManager code but it was requested to live in a new class. This was mirrored in the nativeRenderManager_spec test.

File: src/nativeAssetManager.js
Comment (mmoschovas) [reply src/nativeAssetManager.js]: Updated

File: src/nativeRenderManager.js
Comment (mmoschovas) [reply src/nativeRenderManager.js:70]: Updated

**Comment by robertrmartinez** [comment]
Ok changes look good to me.

Would like @jaiminpanchal27 to do a review before we merge this.

Specifically making sure the Xander proposal of rendererUrl looks good. since they spec'd that part out.

More than happy to set some time or keep an open dialogue if you want @jaiminpanchal27 

Thanks all!

**Comment by robertrmartinez** [comment]
The associated Prebid.js change has been approved and is ready for merge once this has been approved and merged as well:

https://github.com/prebid/Prebid.js/pull/5411

**Review by robertrmartinez — APPROVED** [review]
Looks good to me

**Review by jsnellbaker — APPROVED** [review]
@mmoschovas 
I think the code here looks okay.  Can you please remove the `.DS_Store` file and the `package-lock.json` updates from the PR though?

**Comment by mmoschovas** [comment]
Removed both updates

**Review by wqi1972 — APPROVED** [review]

File: src/nativeAssetManager.js
Comment (robertrmartinez) [inline src/nativeAssetManager.js:274]: Can you split this `flag` out to a new declaration please.

File: src/nativeAssetManager.js
Comment (robertrmartinez) [inline src/nativeAssetManager.js:276]: Any particular reason to use `!=` vs `!===`?

Do we expect adId to be the same data type in both? Or is is possible one is string and one is number?

File: src/nativeRenderManager.js
Comment (robertrmartinez) [reply src/nativeRenderManager.js:70]: Should have a space after closing bracket befor else and after

```javascript
} else {
```

File: src/nativeAssetManager.js
Comment (robertrmartinez) [inline src/nativeAssetManager.js:218]: Could just do

```javascript
return win.document.body.innerHTML.indexOf(str) !== -1
```


Not sure this function is even being used?

File: src/nativeAssetManager.js
Comment (robertrmartinez) [inline src/nativeAssetManager.js:181]: space after closing brackets

File: src/nativeRenderManager.js
Comment (robertrmartinez) [reply src/nativeRenderManager.js:70]: maybe just run enable eslint in your editor and see what else there is?

**Review by robertrmartinez — CHANGES_REQUESTED** [review]
couple minor things

File: src/nativeAssetManager.js
Comment (mmoschovas) [reply src/nativeAssetManager.js]: Done

File: src/nativeAssetManager.js
Comment (mmoschovas) [reply src/nativeAssetManager.js]: Nope good catch. Updated

File: src/nativeAssetManager.js
Comment (mmoschovas) [reply src/nativeAssetManager.js]: I dont remember this function but cant find a use for it so removing

File: src/nativeAssetManager.js
Comment (mmoschovas) [reply src/nativeAssetManager.js]: Fixed these. Not sure how they didnt update last time around

**Comment by bretg** [comment]
Docs PR https://github.com/prebid/prebid.github.io/pull/2479

**Review by allanjun — APPROVED** [review]

**Comment by jsnellbaker** [comment]
@mmoschovas Can you please take a look at the code conflict in the `gulpfile.js`?  I'm not sure if it's a recent thing, but it needs to be addressed.

Thanks.

**Comment by mmoschovas** [comment]
@jsnellbaker committed merge

**Comment by jsnellbaker** [comment]
@mmoschovas Thanks for the updates and for the overall effort behind this PR.  Based on side conversations, we are good to merge this PR (finally).

**Pull request merged**
